feat: CRP-2823 add developer docs for IBE and timelocks (#5811)
dfinity/portal: docs/building-apps/network-features/vetkeys/timelock-encryption.mdx

### "Dead man's switch"

Journalists or whistleblowers could ensure that compromising information in their possession is automatically published if they were to become incapacitated. They can store the information in a dapp, encrypted using IBE that the dapp automatically and publicly recovers when a certain amount of time passes after it last received an authenticated ping from its owner.

> ### Advanced timelocks

Reviewer: Timelocks here may not be the best fit. What about "Policy-based Decryption" or "IBE with advanced access control"

Author: changed to "Advanced decryption policies"

> Anyone can IBE-encrypt their message using a more complex condition as the identity, while a dapp allows anyone who can satisfy that condition, or when the condition is automatically fulfilled, to obtain the corresponding vetKey. This enables interesting use cases by allowing encryption tied to verifiable future events. For example, one could encrypt a message that becomes decryptable only when the price of a stock rises above or falls below a certain level, or implement information escrow and break-the-glass policies.

Reviewer: The text suggests that the conditions are set upon encryption, but they are really enforced by the dapp. I tried to clarify this, ptal

```suggestion
A dapp can enforce additional, programmable conditions before allowing decryption of an IBE-encrypted message. This makes it possible to tie decryption to verifiable future events or complex policies. For example, a message could be encrypted such that it only becomes decryptable when the price of a stock crosses a certain threshold. Other use cases include information escrow, time-triggered disclosures, or break-glass policies that release data under predefined emergency conditions.
```

## How to use timelock encryption?

> We can categorize timelock encryption into two categories: 1. scenario where the canister decrypts, and 2. scenario where the user decrypts.
> The latter is essentially the same as the [IBE](/docs/building-apps/network-features/vetkeys/identity-based-encryption), only with additional conditions on when the key can be obtained by the authorized users.

Reviewer: Good idea on highlighting the two scenarios. Maybe we should expand a little

```suggestion
Timelock encryption can be categorized into two categories depending on whether the result of decryption should be publicly accessible or not: 

1. User-side decryption: The user performs the decryption, ensuring the plaintext remains private.This is suitable for use cases such as delayed private messages or private-data release.
2. Canister-side decryption: The canister performs the decryption to reveal the plaintext. Since canisters do not currently support private or confidential execution, any data they decrypt should be treated as publicly visible. This model is suitable for use cases like time-delayed announcements, secret-bid auctions, or public lotteries, where the decrypted information is meant to be accessible to all users. 

The first category is essentially the same as the [IBE](/docs/building-apps/network-features/vetkeys/identity-based-encryption), with additional restrictions on when the decryption key can be obtained by the authorized users.
```

This can be as simple as checking if enough time has passed and either returning the key or an error to the user.

> Therefore, let's focus on getting the first case to work.

Reviewer: ```suggestion
Therefore, we focus on explaining how to integrate the second category of timelock encryption into a dapp. 
```

> The code in the descriptions below is taken from the [secret-bid auction example](https://github.com/dfinity/vetkeys/tree/3b1de5d546dcc32daaf22b21baaa2b220bc857f3/examples/basic_timelock_ibe), where any user can create an auction lot with a deadline and other users can create IBE-encrypted bids for the open lots.
> The encrypted bids are decrypted by the canister once the deadline has passed and the lot was closed.
> From that moment on, the bids become public and the canister announces the winner of the lot.

Reviewer: Maybe let's expand a bit the description of the secret bid auction, so that it is easier to relate to the steps below. 

```suggestion
The code in the descriptions below is taken from the [secret-bid auction example](https://github.com/dfinity/vetkeys/tree/3b1de5d546dcc32daaf22b21baaa2b220bc857f3/examples/basic_timelock_ibe), which works as follows: 

1. A single (auction?) public key is used by all the users across all auctions. 
2. Any user can create a new auction lot by specifying a unique lot identifier and a bidding deadline. 
3. Users can IBE-encrypt their bids using the public key and the lot identifier.
4. Once the deadline is passed, the auction canister retrieves the decryption key, decrypts all the submitted bids, and determine the highest. From that moment on, the bids become public and the canister announces the winner.
```

> ### Step 1: Obtain IBE public key

Reviewer: Similar to the comments for IBE, I think it would be nice to explain the choice of parameters, like context and input. 

We could also explain why canister_id is set to none. But I don't know if it complicates more than helping

Expose a canister endpoint that allows to fetch the IBE public key of the canister.

<AdornedTabs groupId="languages">

<TabItem value="rs" label="Rust" default>

```rust no-repl reference

> https://github.com/dfinity/vetkeys/blob/3b1de5d546dcc32daaf22b21baaa2b220bc857f3/examples/basic_timelock_ibe/backend/src/lib.rs#L88-L109

Reviewer: I think the endpoint of the example should refer to the dapp, rather than the primitive. E.g. the name of the endpoint could be `get_auction_public_key` instead of `get_root_ibe_public_key`.

Author: I now tried to expose as less as possible information about the auction, so maybe this is not very relevant anymore. But we still mention bids and lots in step 3's referenced code because without any context the time check would be hard to explain. So maybe we could rename the endpoint? Not sure, IMHO it's not too bad right now. But we can discuss.

```

</TabItem>
</AdornedTabs>

To obtain the public key in the frontend, you now only need to make a call to this endpoint, e.g., like it is done [here](https://github.com/dfinity/vetkeys/blob/3b1de5d546dcc32daaf22b21baaa2b220bc857f3/examples/basic_timelock_ibe/frontend/src/main.ts#L53).
The public key needs to only be fetched once and it can be used to encrypt data for multiple recipients by deriving corresponding subkeys locally.

### Step 2: Encrypt data in the frontend

Use the [frontend `@dfinity/vetkeys` library](https://dfinity.github.io/vetkeys/classes/_dfinity_vetkeys.IbeCiphertext.html#encrypt) to encrypt data and store it in the canister.
The `lotId` variable is the identity that we encrypt to to place a bid with the `amount` specified by the user.

<AdornedTabs groupId="languages">

<TabItem value="ts" label="TypeScript" default>

```ts no-repl reference
https://github.com/dfinity/vetkeys/blob/3b1de5d546dcc32daaf22b21baaa2b220bc857f3/examples/basic_timelock_ibe/frontend/src/main.ts#L406-L429
```

</TabItem>
</AdornedTabs>

> ### Step 3: Decrypt data in the backend once conditions are met

Reviewer: It seems to me that the time check is it somewhat important to show. Is it something that could be added here?

Author: I restructured the example code quite significantly, and now the time check is also shown in the referenced code.

2. Canister-side decryption: The canister performs the decryption to reveal the plaintext. Since canisters do not currently support private or confidential execution, any data they decrypt should be treated as publicly visible. This model is suitable for use cases like time-delayed announcements, secret-bid auctions, or public lotteries, where the decrypted information is meant to be accessible to all users. 

The first category is essentially the same as the [IBE](/docs/building-apps/network-features/vetkeys/identity-based-encryption), with additional restrictions on when the decryption key can be obtained by the authorized users.
This can be as simple as checking if enough time has passed and either returning the key or an error to the user.
Therefore, we focus on explaining how to integrate the second category of timelock encryption into a dapp. 

The code in the descriptions below is taken from the [secret-bid auction example](https://github.com/dfinity/vetkeys/tree/3b1de5d546dcc32daaf22b21baaa2b220bc857f3/examples/basic_timelock_ibe), which works as follows: 

> 1. A single (auction?) public key is used by all the users across all auctions. 

Author: No, I just included what Andrea wrote in the review and forgot to change it. I actually think that "(auction) public key" doesn't sound too bad. Changed to that.

---
keywords: [advanced, concept, random, randomness, onchain randomness, raw_rand]
---

import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";
import TabItem from "@theme/TabItem";
import useBaseUrl from "@docusaurus/useBaseUrl";
import { AdornedTabs } from "/src/components/Tabs/AdornedTabs";
import { AdornedTab } from "/src/components/Tabs/AdornedTab";
import { BetaChip } from "/src/components/Chip/BetaChip";

# Timelock encryption

<MarkdownChipRow labels={["Advanced", "vetKeys", "Identity-based encryption", "Timelock"]} />

> A special variant of [identity-based encryption (IBE)](/docs/building-apps/network-features/vetkeys/identity-based-encryption) is called timelock encryption.
> It allows a sender to encrypt to a specific timestamp, ensuring that the recipient can only decrypt after the specific time has passed. Canisters can enforce this time-based access control by requesting threshold decryption of a ciphertext only after a predetermined expiry time, keeping sensitive information secret until the appropriate moment.
> This makes it possible to implement time-sensitive applications, like secret-bid auctions, time-locked documents, dead-man switches, and delayed reveal NFTs.
> Timelock encryption also serves as a key building block for protecting against maximal extractable value (MEV). By keeping transaction details confidential until after block inclusion, it prevents adversaries from doing front-running and reordering of transactions.

Reviewer 2: ```suggestion
A special variant of [identity-based encryption (IBE)](/docs/building-apps/network-features/vetkeys/identity-based-encryption) is called timelock encryption.  
It enables a sender to encrypt data to a specific timestamp, ensuring that the recipient can only decrypt it after the specified time has passed.

Canisters enforce this time-based access control by initiating threshold decryption of a cipher text only after a predefined expiry time. This mechanism ensures that sensitive information remains secret until the appropriate moment.

Timelock encryption enables a wide range of time-sensitive applications, such as:

- Secret-bid auctions  
- Time-locked documents  
- Dead-man switches  
- Delayed-reveal NFTs

It also plays a key role in protecting against maximal extractable value (MEV). By keeping transaction details confidential until after block inclusion, it prevents adversaries from front-running or reordering transactions.
```

### "Dead man's switch"

Journalists or whistleblowers could ensure that compromising information in their possession is automatically published if they were to become incapacitated. The information is IBE-encrypted and stored it in a dapp, which is configured to automatically decrypt it and publish if it doesn't receive a periodic ping from the autorized user.

### Advanced decryption policies

A dapp can enforce additional, programmable conditions before allowing decryption of an IBE-encrypted message. This makes it possible to tie decryption to verifiable future events or complex policies. For example, a message could be encrypted such that it only becomes decryptable when the price of a stock crosses a certain threshold. Other use cases include information escrow, time-triggered disclosures, or break-glass policies that release data under predefined emergency conditions.

> ## How to use timelock encryption?

Reviewer 2: ```suggestion
## How to use timelock encryption
```

Timelock encryption can be categorized into two categories depending on whether the result of decryption should be publicly accessible or not: 

1. User-side decryption: The user performs the decryption, ensuring the plaintext remains private. This is suitable for use cases such as delayed private messages or private-data release.
2. Canister-side decryption: The canister performs the decryption to reveal the plaintext. Since canisters do not currently support private or confidential execution, any data they decrypt should be treated as publicly visible. This model is suitable for use cases like time-delayed announcements, secret-bid auctions, or public lotteries, where the decrypted information is meant to be accessible to all users. 

The first category is essentially the same as the [IBE](/docs/building-apps/network-features/vetkeys/identity-based-encryption), with additional restrictions on when the decryption key can be obtained by the authorized users.
This can be as simple as checking if enough time has passed and either returning the key or an error to the user.

> Therefore, we focus on explaining how to integrate the second category of timelock encryption into a dapp. 

Reviewer 2: ```suggestion
Therefore, this documentation explains how to integrate the second category of timelock encryption into a dapp. 
```

> The code in the descriptions below is taken from the [secret-bid auction example](https://github.com/dfinity/vetkeys/tree/dd255c8fa1ec0356f9448f1728ed4d6a5b736308/examples/basic_timelock_ibe), which works as follows: 

Reviewer 2: ```suggestion
The code featured below is taken from the [secret-bid auction example](https://github.com/dfinity/vetkeys/tree/dd255c8fa1ec0356f9448f1728ed4d6a5b736308/examples/basic_timelock_ibe), which works as follows: 
```

1. A single (auction) public key is used by all the users across all auctions. 
2. Any user can create a new auction lot by specifying a unique lot identifier and a bidding deadline. 
3. Users can IBE-encrypt their bids using the public key and the lot identifier.
4. Once the deadline has passed, the auction canister retrieves the decryption key, decrypts all the submitted bids, and determine the highest. From that moment on, the bids become public and the canister announces the winner.

> ### Step 1: Allow to obtain IBE public key from the backend

Reviewer 2: ```suggestion
### Step 1: Obtain IBE public key from the backend
```

> Expose a canister endpoint that allows to fetch the IBE public key of the canister.

Reviewer 2: ```suggestion
Expose a canister endpoint that can fetch the IBE public key of the canister.
```

<AdornedTabs groupId="languages">

<TabItem value="rs" label="Rust" default>

> ```rust no-repl reference

Reviewer 2: ```suggestion
```rust reference
```

https://github.com/dfinity/vetkeys/blob/dd255c8fa1ec0356f9448f1728ed4d6a5b736308/examples/basic_timelock_ibe/backend/src/lib.rs#L100-L121
```

</TabItem>
</AdornedTabs>

> The public key needs to only be fetched once and it can be used to encrypt data for multiple recipients by deriving corresponding subkeys locally.
> It is a good practice to use a domain separator in `context` to split IBE from other eventual uses of vetKeys the canister may do.
> A domain separator can be any arbitrary unique byte string.

Reviewer 2: ```suggestion
The public key only needs to be fetched once and can then be reused to encrypt data for multiple recipients by deriving the corresponding subkeys locally. To ensure cryptographic separation and prevent cross-protocol interference, it's good practice to include a domain separator in the `context` when using IBE. This distinguishes IBE usage from any other potential applications of vetKeys that the canister may support. The domain separator can be any unique and arbitrary byte string.
```

> ### Step 2: Obtain IBE public key and encrypt data in the frontend

Reviewer 2: ```suggestion
### Step 2: Obtain the IBE public key and encrypt data in the frontend
```

Use the [frontend `@dfinity/vetkeys` library](https://dfinity.github.io/vetkeys/classes/_dfinity_vetkeys.IbeCiphertext.html#encrypt) to encrypt data and store it in the canister.

> To encrypt bids in the example, the bidder uses the auction lot id as the encryption identity (specified in the `input` variable of the `vetkd_derive_key` arguments).

Reviewer 2: ```suggestion
To encrypt bids in the example, the bidder uses the auction lot ID as the encryption identity (specified in the `input` variable of the `vetkd_derive_key` arguments).
```

The alternative to encrypting to unique ids is encrypting to timestamps that define when the data should be decrypted, which may reduce the number of the required calls to the vetKD API for deriving decryption keys.

<AdornedTabs groupId="languages">

<TabItem value="ts" label="TypeScript" default>

> ```ts no-repl reference

Reviewer 2: ```suggestion
```ts reference
```

https://github.com/dfinity/vetkeys/blob/dd255c8fa1ec0356f9448f1728ed4d6a5b736308/examples/basic_timelock_ibe/frontend/src/main.ts#L185-L207
```

</TabItem>
</AdornedTabs>

### Step 3: Decrypt data in the backend once conditions are met

> The condition for decryption in the secret bid auction example is that the current time should be after the deadline time.
> In other example, the condition may change with time (e.g., in dead man's switch the deadline may be extended) or be more complex and depend on external factors, such as allowing decryption only when the price of a stock gets higher than some threshold, determined via HTTPS outcalls.

Reviewer 2: ```suggestion
In the secret-bid auction example, the condition for decryption is straightforward: the current time must be after the specified deadline.

In other scenarios, the decryption condition may evolve over time or depend on external factors. For example:

- In a dead man's switch, the deadline may be extended if the user checks in periodically.
- In a conditional release based on market data, decryption may only be allowed when the price of a stock exceeds a certain threshold, determined via HTTPS outcalls.

These flexible conditions allow timelock encryption to support a wide range of dynamic and context-aware use cases.
```

In the code below, the time condition is checked by simply comparing the stored timestamp with the timestamp of the current time obtained using the `ic-cdk` crate.
There, a timer is invoked periodically to check if there are any lots to close which triggers the lot closure and bid decryption.

> The latter is implemented in the `decrypt_bids` function showed below.

Reviewer 2: ```suggestion
The latter is implemented in the `decrypt_bids` function shown below.
```

<AdornedTabs groupId="languages">

<TabItem value="rs" label="Rust" default>

> ```rust no-repl reference

Reviewer 2: ```suggestion
```rust reference
```
